core/src/main/scala/org/apache/spark/sql/store/StoreUtils.scala

```diff
@@ -98,25 +101,74 @@ object StoreUtils extends Logging {
 
   val SHADOW_COLUMN = s"$SHADOW_COLUMN_NAME bigint generated always as identity"
 
-  def getPartitionsPartitionedTable(sc: SparkContext,
-      region: PartitionedRegion): Array[Partition] = {
-
-    val numPartitions = region.getTotalNumberOfBuckets
-    val partitions = new Array[Partition](numPartitions)
+  def lookupName(tableName: String, schema: String): String = {
+    val lookupName = {
+      if (tableName.indexOf('.') <= 0) {
+        schema + '.' + tableName
+      } else tableName
+    }
+    lookupName
+  }
 
-    for (p <- 0 until numPartitions) {
-      val distMembers = region.getRegionAdvisor.getBucketOwners(p).asScala
-      val prefNodes = distMembers.map(
-        m => SnappyContext.storeToBlockMap.get(m.toString)
-      ).filter(_.isDefined)
-      val prefNodeSeq = prefNodes.map(_.get).toSeq
-      partitions(p) = new MultiExecutorLocalPartition(p, prefNodeSeq)
+  def getPartitionsPartitionedTable(sc: SparkContext,
+                                    region: PartitionedRegion): Array[Partition] = {
+
+    val numBuckets = region.getTotalNumberOfBuckets
+    val serverToBuckets = new mutable.HashMap[InternalDistributedMember, mutable.HashSet[Int]]()
+    var totalBuckets = new mutable.HashSet[Int]()
+    for (p <- 0 until numBuckets) {
+      val owner = region.getBucketPrimary(p)
+      val bucketSet = {
+        if (serverToBuckets.contains(owner)) serverToBuckets.get(owner).get
+        else new mutable.HashSet[Int]()
+      }
+      bucketSet += p
+      totalBuckets += p
+      serverToBuckets put(owner, bucketSet)
+    }
+    val numCores = Runtime.getRuntime.availableProcessors()
+    val numServers = GemFireXDUtils.getGfxdAdvisor.adviseDataStores(null).size()
```

Comment:
We have assumed that, the cluster will be homogeneous with respect number of cores on each server. I will check if there is a way to handle heterogeneous case also.